linera-views/src/backends/dual.rs

```diff
@@ -0,0 +1,496 @@
+            ),
+            StoreInUse::Second => DualStoreKeyValues::Second(
+                self.second_store
+                    .find_key_values_by_prefix(key_prefix)
+                    .await
+                    .map_err(DualStoreError::Second)?,
+            ),
+        };
+        Ok(result)
+    }
+}
+
+impl<S1, S2, A> WritableKeyValueStore for DualStore<S1, S2, A>
+where
+    S1: WritableKeyValueStore + WithError + Send + Sync,
+    S2: WritableKeyValueStore + WithError + Send + Sync,
+    A: Send + Sync,
+{
+    const MAX_VALUE_SIZE: usize = usize::MAX;
+
+    async fn write_batch(&self, batch: Batch) -> Result<(), Self::Error> {
+        match self.store_in_use {
+            StoreInUse::First => self
+                .first_store
+                .write_batch(batch)
+                .await
+                .map_err(DualStoreError::First)?,
+            StoreInUse::Second => self
+                .second_store
+                .write_batch(batch)
+                .await
+                .map_err(DualStoreError::Second)?,
+        }
+        Ok(())
+    }
+
+    async fn clear_journal(&self) -> Result<(), Self::Error> {
+        match self.store_in_use {
+            StoreInUse::First => self
+                .first_store
+                .clear_journal()
+                .await
+                .map_err(DualStoreError::First)?,
+            StoreInUse::Second => self
+                .second_store
+                .clear_journal()
+                .await
+                .map_err(DualStoreError::Second)?,
+        }
+        Ok(())
+    }
+}
+
+impl<S1, S2, A> AdminKeyValueStore for DualStore<S1, S2, A>
+where
+    S1: AdminKeyValueStore + Send + Sync,
+    S2: AdminKeyValueStore + Send + Sync,
+    A: DualStoreRootKeyAssignment + Send + Sync,
+{
+    type Config = DualStoreConfig<S1::Config, S2::Config>;
+
+    async fn new_test_config() -> Result<Self::Config, Self::Error> {
+        let first_config = S1::new_test_config().await.map_err(DualStoreError::First)?;
+        let second_config = S2::new_test_config()
+            .await
+            .map_err(DualStoreError::Second)?;
+        Ok(DualStoreConfig {
+            first_config,
+            second_config,
+        })
+    }
+
+    async fn connect(
+        config: &Self::Config,
+        namespace: &str,
+        root_key: &[u8],
+    ) -> Result<Self, Self::Error> {
+        let first_store = S1::connect(&config.first_config, namespace, root_key)
+            .await
+            .map_err(DualStoreError::First)?;
+        let second_store = S2::connect(&config.second_config, namespace, root_key)
+            .await
+            .map_err(DualStoreError::Second)?;
+        let store_in_use = A::assigned_store(root_key)?;
+        Ok(Self {
+            first_store,
+            second_store,
+            store_in_use,
+            _marker: std::marker::PhantomData,
+        })
+    }
+
+    fn clone_with_root_key(&self, root_key: &[u8]) -> Result<Self, Self::Error> {
```

Comment:
Note: Because this function is not async, we're forced to open connections to all stores (instead of the one in use) in `connect`.